RZDataBinding/NSObject+RZDataBinding.m

```diff
@@ -412,28 +468,202 @@ - (instancetype)init
+{
+    RZDBNotification *notification = [[self alloc] init];
+    notification.target = observer.target;
+    notification.action = observer.action;
+    notification.queue = observer.callbackQueue;
+    notification.changeDict = change;
+
+    return notification;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    @synchronized (self) {
+        BOOL equal = [super isEqual:object];
+
+        if ( !equal && [object isKindOfClass:[RZDBNotification class]] ) {
+            RZDBNotification *other = (RZDBNotification *)object;
+
+            // Do the callback queues need to be equal?
+            equal = (self.target == other.target) &&
+                    (self.action == other.action) &&
+                    (self.changeDict == other.changeDict ||
+                     (self.changeDict[kRZDBChangeKeyObject] == other.changeDict[kRZDBChangeKeyObject] &&
+                      [self.changeDict[kRZDBChangeKeyKeyPath] isEqualToString:other.changeDict[kRZDBChangeKeyKeyPath]]
+                     )
+                    );
+        }
+
+        return equal;
+    }
+}
+
+- (void)send
+{
+    @synchronized (self) {
+        if ( self.target != nil && self.action != NULL ) {
+            void (^notificationBlock)();
+
+            if ( self.changeDict != nil ) {
+                notificationBlock = ^{
+                    ((void(*)(id, SEL, NSDictionary *))objc_msgSend)(self.target, self.action, [self.changeDict copy]);
+                };
+            }
+            else {
+                notificationBlock = ^{
+                    ((void(*)(id, SEL))objc_msgSend)(self.target, self.action);
+                };
+            }
+
+            if ( self.queue == nil ) {
+                notificationBlock();
+            }
+            else {
+                dispatch_async(self.queue, notificationBlock);
+            }
+        }
+    }
+}
+
+@end
+
+#pragma mark - RZDBTransaction implementation
+
+@implementation RZDBTransaction
+
++ (void)begin
+{
+    RZDBTransaction *current = [self currentTransaction];
+
+    if ( current == nil ) {
+        [self setCurrentTransaction:[[self alloc] init]];
+    }
+    else {
+        current.beginCount++;
+    }
+}
+
++ (void)commit
+{
+    RZDBTransaction *current = [self currentTransaction];
+    current.beginCount--;
+
+    if ( current.beginCount == 0 ) {
+        [self setCurrentTransaction:nil];
+        [[current.notifications copy] enumerateObjectsUsingBlock:^(RZDBNotification *notification, NSUInteger idx, BOOL *stop) {
+            [notification send];
+        }];
+    }
+}
+
++ (void)transactionWithBlock:(void (^)())transactionBlock
+{
+    NSParameterAssert(transactionBlock);
+
+    [RZDBTransaction begin];
+    transactionBlock();
+    [RZDBTransaction commit];
+}
+
+#pragma mark - private methods
+
++ (RZDBTransaction *)currentTransaction
+{
+    return [NSThread currentThread].threadDictionary[kRZDBTransactionStorageKey];
```

Comment:
I still don't like dot notation for anything besides properties in Objc. Just a preference.